resource/aws_security_group_rule: Add import functionality for security group rules

YakDriver opened the pull request:
Fixes #2895

Changes proposed in this pull request:

* Add functionality to the existing resource `aws_security_group_rule` to allow for importing rules
* Add importing and import verification to the acceptance tests
* Add import documentation

## How it works

### Import string/ID

The import string/ID can be complex but manageable. It follows this format:

`SECURITYGROUPID_TYPE_PROTOCOL_FROMPORT_TOPORT_SOURCE[_SOURCE]*`

These are valid examples of import strings:
* `sg-09a093729ef9382a6_ingress_tcp_8000_8000_10.0.3.0/24`
* `sg-09a093729ef9382a6_ingress_92_0_65535_10.0.3.0/24_10.0.4.0/24` (multiple CIDR blocks)
* `sg-09a093729ef9382a6_egress_tcp_100_8000_10.0.3.0/24` (port range)
* `sg-09a093729ef9382a6_egress_tcp_8000_8000_pl-34800000` (prefix list id destination)
* `sg-09a093729ef9382a6_ingress_all_0_65535_sg-08123412342323`
* `sg-09a093729ef9382a6_ingress_tcp_100_121_10.1.0.0/16_2001:db8::/48_10.2.0.0/16_2002:db8::/48` (mixture of IPv4 and IPv6 CIDRs)
* `sg-09a093729ef9382a6_ingress_tcp_0_65535_self_2001:db8::/48` (self + IPv6 CIDR block)
* `sg-09a093729ef9382a6_ingress_92_0_65535_self` (if protocol is not TCP/UDP/ICMP/ALL then it must be a protocol number)

### All tricky functionality 

This captures all security group rule functionality including the tricky bits:
* as many CIDR blocks as someone wants to type
* self
* `all` protocols
* `all` ports
* port ranges
* multiple IPv4/6 CIDR blocks, prefix list IDs
* source/destination security groups

### Flexible: Config / import ID don't have to match

AWS is not particular about needing to modify an entire rule. Thus, you can partially import (i.e., some CIDR blocks but not others) and then update those within Terraform. AWS gracefully adjusts rules accordingly.

## Acceptance tests:

I added imports/state verification to 17 of 20 existing tests. Import didn't make sense for the other 3 -- 2 test error conditions and 1 tests a race condition.

Output from acceptance testing:

```console
$ make testacc TESTARGS='-run=TestAccAWSSecurityGroupRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSSecurityGroupRule_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSSecurityGroupRule_Ingress_VPC
--- PASS: TestAccAWSSecurityGroupRule_Ingress_VPC (21.20s)
=== RUN   TestAccAWSSecurityGroupRule_Ingress_Protocol
--- PASS: TestAccAWSSecurityGroupRule_Ingress_Protocol (35.63s)
=== RUN   TestAccAWSSecurityGroupRule_Ingress_Ipv6
--- PASS: TestAccAWSSecurityGroupRule_Ingress_Ipv6 (35.64s)
=== RUN   TestAccAWSSecurityGroupRule_Ingress_Classic
--- PASS: TestAccAWSSecurityGroupRule_Ingress_Classic (20.46s)
=== RUN   TestAccAWSSecurityGroupRule_MultiIngress
--- PASS: TestAccAWSSecurityGroupRule_MultiIngress (23.22s)
=== RUN   TestAccAWSSecurityGroupRule_Egress
--- PASS: TestAccAWSSecurityGroupRule_Egress (18.58s)
=== RUN   TestAccAWSSecurityGroupRule_SelfReference
--- PASS: TestAccAWSSecurityGroupRule_SelfReference (35.63s)
=== RUN   TestAccAWSSecurityGroupRule_ExpectInvalidTypeError
--- PASS: TestAccAWSSecurityGroupRule_ExpectInvalidTypeError (0.85s)
=== RUN   TestAccAWSSecurityGroupRule_ExpectInvalidCIDR
--- PASS: TestAccAWSSecurityGroupRule_ExpectInvalidCIDR (0.98s)
=== RUN   TestAccAWSSecurityGroupRule_PartialMatching_basic
--- PASS: TestAccAWSSecurityGroupRule_PartialMatching_basic (41.64s)
=== RUN   TestAccAWSSecurityGroupRule_PartialMatching_Source
--- PASS: TestAccAWSSecurityGroupRule_PartialMatching_Source (37.77s)
=== RUN   TestAccAWSSecurityGroupRule_Issue5310
--- PASS: TestAccAWSSecurityGroupRule_Issue5310 (18.91s)
=== RUN   TestAccAWSSecurityGroupRule_Race
--- PASS: TestAccAWSSecurityGroupRule_Race (318.46s)
=== RUN   TestAccAWSSecurityGroupRule_SelfSource
--- PASS: TestAccAWSSecurityGroupRule_SelfSource (35.23s)
=== RUN   TestAccAWSSecurityGroupRule_PrefixListEgress
--- PASS: TestAccAWSSecurityGroupRule_PrefixListEgress (50.37s)
=== RUN   TestAccAWSSecurityGroupRule_IngressDescription
--- PASS: TestAccAWSSecurityGroupRule_IngressDescription (21.09s)
=== RUN   TestAccAWSSecurityGroupRule_EgressDescription
--- PASS: TestAccAWSSecurityGroupRule_EgressDescription (20.27s)
=== RUN   TestAccAWSSecurityGroupRule_IngressDescription_updates
--- PASS: TestAccAWSSecurityGroupRule_IngressDescription_updates (30.31s)
=== RUN   TestAccAWSSecurityGroupRule_EgressDescription_updates
--- PASS: TestAccAWSSecurityGroupRule_EgressDescription_updates (29.49s)
=== RUN   TestAccAWSSecurityGroupRule_MultiDescription
--- PASS: TestAccAWSSecurityGroupRule_MultiDescription (87.41s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	866.721s
```

YakDriver commented:
@bflad code, acceptance tests, docs ready for review & adjust!

bflad commented on aws/resource_aws_security_group_rule.go:
```diff
@@ -845,3 +872,127 @@ func resourceSecurityGroupRuleDescriptionUpdate(conn *ec2.EC2, d *schema.Resourc
+		return nil, fmt.Errorf(errStr, importStr, "invalid security group ID")
+	}
+
+	if ruleType != "in" && ruleType != "out" {
```
Could you adjust these to `ingress` and `egress` to match the resources and API please?

bflad commented on website/docs/r/security_group_rule.html.markdown:
```diff
@@ -87,3 +87,23 @@ In addition to all arguments above, the following attributes are exported:
+
+## Import
+
+Individual ingress rules can be imported using `SGID_IN_PROTOCOL_PORT_SOURCE[_SOURCE]*`. Egress rules are imported similarly using `SGID_OUT_PROTOCOL_PORT_DESTINATION[_DESTINATION]*`. All parts are required.
```
I'd suggest potential updates to this format, depending on the above, e.g. `SECURITYGROUPID_TYPE_PROTOCOL_FROMPORT_TOPORT_SOURCE[_SOURCE]*` 😄 

It may also be better to spell this out instead and rely on many examples below, e.g.

```
Security Group Rules can be imported using their `security_group_id`, `type`, `protocol`, `from_port`, `to_port`, and source(s) (e.g. `cidr_block`) separated by underscores (`_`).
```

YakDriver replied:
Updated throughout.

YakDriver commented on aws/resource_aws_security_group_rule.go:
```diff
@@ -845,3 +872,127 @@ func resourceSecurityGroupRuleDescriptionUpdate(conn *ec2.EC2, d *schema.Resourc
+	}
+
+	sgID := importParts[0]
+	ruleType := importParts[1]
```
All the parts go to lower at line 887.

YakDriver commented on aws/resource_aws_security_group_rule.go:
```diff
@@ -845,3 +872,127 @@ func resourceSecurityGroupRuleDescriptionUpdate(conn *ec2.EC2, d *schema.Resourc
+
+	d.Set("protocol", protocolForValue(protocol))
+
+	if portsRaw == "all" {
```
Will do.

YakDriver replied:
This is implemented throughout code, tests, doc.

YakDriver commented:
@bflad I made the changes you requested, added more examples to the docs, updated the examples above in this PR, and have re-run the acceptance tests:

```console
$ make testacc TESTARGS='-run=TestAccAWSSecurityGroupRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSSecurityGroupRule_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSSecurityGroupRule_Ingress_VPC
--- PASS: TestAccAWSSecurityGroupRule_Ingress_VPC (21.82s)
=== RUN   TestAccAWSSecurityGroupRule_Ingress_Protocol
--- PASS: TestAccAWSSecurityGroupRule_Ingress_Protocol (35.72s)
=== RUN   TestAccAWSSecurityGroupRule_Ingress_Ipv6
--- PASS: TestAccAWSSecurityGroupRule_Ingress_Ipv6 (34.94s)
=== RUN   TestAccAWSSecurityGroupRule_Ingress_Classic
--- PASS: TestAccAWSSecurityGroupRule_Ingress_Classic (20.69s)
=== RUN   TestAccAWSSecurityGroupRule_MultiIngress
--- PASS: TestAccAWSSecurityGroupRule_MultiIngress (23.07s)
=== RUN   TestAccAWSSecurityGroupRule_Egress
--- PASS: TestAccAWSSecurityGroupRule_Egress (20.14s)
=== RUN   TestAccAWSSecurityGroupRule_SelfReference
--- PASS: TestAccAWSSecurityGroupRule_SelfReference (36.87s)
=== RUN   TestAccAWSSecurityGroupRule_ExpectInvalidTypeError
--- PASS: TestAccAWSSecurityGroupRule_ExpectInvalidTypeError (0.92s)
=== RUN   TestAccAWSSecurityGroupRule_ExpectInvalidCIDR
--- PASS: TestAccAWSSecurityGroupRule_ExpectInvalidCIDR (0.87s)
=== RUN   TestAccAWSSecurityGroupRule_PartialMatching_basic
--- PASS: TestAccAWSSecurityGroupRule_PartialMatching_basic (42.00s)
=== RUN   TestAccAWSSecurityGroupRule_PartialMatching_Source
--- PASS: TestAccAWSSecurityGroupRule_PartialMatching_Source (41.27s)
=== RUN   TestAccAWSSecurityGroupRule_Issue5310
--- PASS: TestAccAWSSecurityGroupRule_Issue5310 (20.27s)
=== RUN   TestAccAWSSecurityGroupRule_Race
--- PASS: TestAccAWSSecurityGroupRule_Race (323.01s)
=== RUN   TestAccAWSSecurityGroupRule_SelfSource
--- PASS: TestAccAWSSecurityGroupRule_SelfSource (35.96s)
=== RUN   TestAccAWSSecurityGroupRule_PrefixListEgress
--- PASS: TestAccAWSSecurityGroupRule_PrefixListEgress (48.83s)
=== RUN   TestAccAWSSecurityGroupRule_IngressDescription
--- PASS: TestAccAWSSecurityGroupRule_IngressDescription (20.90s)
=== RUN   TestAccAWSSecurityGroupRule_EgressDescription
--- PASS: TestAccAWSSecurityGroupRule_EgressDescription (20.36s)
=== RUN   TestAccAWSSecurityGroupRule_IngressDescription_updates
--- PASS: TestAccAWSSecurityGroupRule_IngressDescription_updates (29.80s)
=== RUN   TestAccAWSSecurityGroupRule_EgressDescription_updates
--- PASS: TestAccAWSSecurityGroupRule_EgressDescription_updates (29.29s)
=== RUN   TestAccAWSSecurityGroupRule_MultiDescription
--- PASS: TestAccAWSSecurityGroupRule_MultiDescription (126.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	933.525s
```

bflad commented on aws/resource_aws_security_group_rule.go:
```diff
@@ -23,6 +24,25 @@ func resourceAwsSecurityGroupRule() *schema.Resource {
+				if err := populateSecurityGroupRuleFromImport(d, importParts); err != nil {
+					return nil, err
+				}
+				err = resourceAwsSecurityGroupRuleRead(d, meta)
```
The `Read` function is automatically called, so this can be removed and simply: `return []*schema.ResourceData{d}, nil` 😄

YakDriver replied:
👍

bflad commented:
This has been released in [version 1.40.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1400-october-10-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
